### ISSUE TITLE
AppImage does not run on Ubuntu LTS (and older systems)

### ISSUE THREAD

**Issue by TheAssassin** [jonaski/strawberry#101]
Seems like you're building your AppImage on a _very recent_ system (it's linked against glibc 2.29, not even Debian unstable/sid has that!). This basically means that noone is able to use your AppImage, since they don't have such a new version. Please see https://docs.appimage.org/introduction/concepts.html#build-on-old-systems-run-on-newer-systems.

Would it be possible you build the AppImage e.g., on xenial (or if that is impossible, at least bionic; but xenial is clearly preferred)? This will greatly increase your target audience for the AppImage, and will help spread the word about Strawberry Music Player. 

I just checked, and any sort of rolling release distro like Tumbleweed is generally a horrible choice for making any sort of packages. Your app seems to build on xenial just fine, so it's basically a matter of switching the base image for the build.

**Comment by 5ay3h** [comment]
Also on latest Debian testing: `4.19.0-4-amd64 #1 SMP Debian 4.19.28-2 (2019-03-15) x86_64 GNU/Linux`

**Comment by TheAssassin** [comment]
@5ay3h yup:

> Seems like you're building your AppImage on a very recent system (it's linked against glibc 2.29, not even Debian unstable/sid has that!).

That's the issue here. Please read https://docs.appimage.org/introduction/concepts.html#build-on-old-systems-run-on-newer-systems.

**Comment by TheAssassin** [comment]
Preview of a new tool I'm working on:

```
> appimagelint Strawberry-0.5.3-8-gd1c2188.AppImage 
appimagelint.cli[22965] [INFO] Downloading up-to-date runtime from GitHub
appimagelint.cli[22965] [INFO] Checking AppImage Strawberry-0.5.3-8-gd1c2188.AppImage
appimagelint.cli[22965] [INFO] Running check "GNU libc ABI check"
appimagelint.glibc_abi_check[22965] [INFO] detected required glibc version for runtime: 2.3.3
appimagelint.glibc_abi_check[22965] [INFO] detected required glibc version for payload: 2.25
appimagelint.glibc_abi_check[22965] [INFO] [✖] AppImage can run on Debian oldstable (jessie)
appimagelint.glibc_abi_check[22965] [INFO] [✖] AppImage can run on Debian stable (stretch)
appimagelint.glibc_abi_check[22965] [INFO] [✔] AppImage can run on Debian testing (buster)
appimagelint.glibc_abi_check[22965] [INFO] [✔] AppImage can run on Debian unstable (sid)
appimagelint.glibc_abi_check[22965] [INFO] [✖] AppImage can run on Ubuntu trusty
appimagelint.glibc_abi_check[22965] [INFO] [✖] AppImage can run on Ubuntu xenial
appimagelint.glibc_abi_check[22965] [INFO] [✔] AppImage can run on Ubuntu bionic
appimagelint.glibc_abi_check[22965] [INFO] [✔] AppImage can run on Ubuntu cosmic
appimagelint.glibc_abi_check[22965] [INFO] [✔] AppImage can run on Ubuntu disco
```

Much better, thanks! Any reason not to "just do it" on xenial (or maybe Debian stable), though? I'm curious. It seems you can build the player on xenial normally, too.

**Comment by jonaski** [comment]
The AppImages were broken, not only this but gstreamer plugins were missing too. I've removed them from releases and the builds site. But I didn't publish them on those various sites, someone else did.
I never liked debian based distros, so I tend to stay away from them, also Xenial only has Qt 5.5, and I prefer to use Qt 5.6 or later.
I think maybe centos could work well, it may seem like an odd choice but it's LTS, it got old core libraries, glibc 2.17, but Qt 5.9 and gstreamer 1.14 is available through a different repository. The second choice is mageia glibc 2.22. I tested appimage bulit on centos on xenial and it runs fine, except the gstreamer plugins which needs to be fixed too.
Would that work?
I think the one you downloaded was built on opensuse leap 43.2, I struggled with broken gstreamer plugins on centos and mageia so I was just testing it, but I think it was because the deploy tool does not find gstreamer plugins and picks up my systems plugins instead. Opensuse releases have too new packages and are too shortlived to be suitable.
Is there a tool to deploy gstreamer plugins? I only need audio plugins, not video.

**Comment by jonaski** [comment]
I'm not sure how to deploy the gstreamer plugins.
It keeps looking in /usr/lib64/gstreamer-1.0/ for the plugins, which will break unless they are the except same version and linked with the same symbols.

**Comment by TheAssassin** [comment]
Let me check if there's a "nice trick" known for this. Last time I was working on gstreamer related stuff was 2017, though, and I can't remember that there was "that one nice trick" that made things magically work.

**Comment by TheAssassin** [comment]
Try to add an `AppRun` shell script that exports the variable `GST_PLUGIN_SYSTEM_PATH` to where the files really are (something like `export GST_PLUGIN_SYSTEM_PATH=$(readlink -f $(dirname "$0")/usr/lib/gstreamer`).

Source: https://gstreamer.freedesktop.org/data/doc/gstreamer/head/gstreamer/html/gst-running.html

**Comment by jonaski** [comment]
Can you try this now:
https://builds.strawbs.net/appimage/Strawberry-0.5.3.AppImage
It is built on CentOs and has gstreamer plugins.

One minor issue is that it does not follow Qt styles set in qt5ct (not sure what files are needed for this to be fixed).

I also removed the appdata file because of this issue:
```
-- Running output plugin: appimage -- 
Running process: /home/jonas/Projects/strawberry/strawberry-appimage/build/linuxdeploy-plugin-appimage-x86_64.AppImage --appdir AppDir
[appimage/stdout] Found appimagetool: /tmp/.mount_linuxdg1ZL6g/usr/bin/appimagetool
[appimage/stdout] 
[appimage/stderr] Running command: /tmp/.mount_linuxdg1ZL6g/usr/bin/appimagetool "AppDir" "-g"
[appimage/stderr] 
[appimage/stderr] appimagetool, continuous build (commit 2db94af), build 2074 built on 2019-04-06 02:12:40 UTC
[appimage/stderr] NOTE: Using the output of 'git rev-parse --short HEAD' as the version:
[appimage/stderr]       9dded52
[appimage/stderr]       Please set the $VERSION environment variable if this is not intended
[appimage/stderr] Using architecture x86_64
[appimage/stderr] Deleting pre-existing .DirIcon
[appimage/stderr] Creating .DirIcon symlink based on information from desktop file
[appimage/stderr] AppStream upstream metadata found in usr/share/metainfo/org.strawbs.strawberry.appdata.xml
[appimage/stdout] /home/jonas/Projects/strawberry/strawberry-appimage/build/AppDir should be packaged as Strawberry-9dded52-x86_64.AppImage
[appimage/stdout] Trying to validate AppStream information with the appstreamcli tool
[appimage/stdout] In case of issues, please refer to https://github.com/ximion/appstream
[appimage/stdout] Validation was successful.
[appimage/stdout] Trying to validate AppStream information with the appstream-util tool
[appimage/stdout] In case of issues, please refer to https://github.com/hughsie/appstream-glib
[appimage/stdout] /home/jonas/Projects/strawberry/strawberry-appimage/build/AppDir/usr/share/metainfo/org.strawbs.strawberry.appdata.xml: FAILED:
[appimage/stdout] • url-not-found         : <screenshot> failed to connect: SSL handshake failed [https://www.strawbs.org/pictures/appdata-screenshot-001.png]
[appimage/stdout] • url-not-found         : <screenshot> failed to connect: SSL handshake failed [https://www.strawbs.org/pictures/appdata-screenshot-002.png]
[appimage/stderr] Validation of files failed
[appimage/stderr] Failed to validate AppStream information with appstream-util
ERROR: Failed to run plugin: appimage 

```

**Comment by TheAssassin** [comment]
The `appstream-cli` version on CentOS is probably quite outdated (or some libraries are outdated and broken).

Qt 5 style stuff is always a bit problematic. Please feel free to open issues about that in linuxdeploy-plugin-qt, we can have a look at that then.

I'm atm extending my tool with support for libstdc++ ABI compatibility. I'll include yours into my test AppImages suite.

**Comment by TheAssassin** [comment]
Alpha test for libstdc++ compatibility:

```
appimagelint.cli[21818] [INFO] Checking AppImage Strawberry-0.5.3.AppImage
appimagelint.cli[21818] [INFO] Running check "GNU libc ABI check"
appimagelint.glibc_abi_check[21818] [INFO] detected required version for runtime: 2.3.3
appimagelint.glibc_abi_check[21818] [INFO] detected required version for payload: 2.9
appimagelint.glibc_abi_check[21818] [INFO] [✔] AppImage can run on Debian oldstable (jessie)
appimagelint.glibc_abi_check[21818] [INFO] [✔] AppImage can run on Debian stable (stretch)
appimagelint.glibc_abi_check[21818] [INFO] [✔] AppImage can run on Debian testing (buster)
appimagelint.glibc_abi_check[21818] [INFO] [✔] AppImage can run on Debian unstable (sid)
appimagelint.glibc_abi_check[21818] [INFO] [✔] AppImage can run on Ubuntu trusty
appimagelint.glibc_abi_check[21818] [INFO] [✔] AppImage can run on Ubuntu xenial
appimagelint.glibc_abi_check[21818] [INFO] [✔] AppImage can run on Ubuntu bionic
appimagelint.glibc_abi_check[21818] [INFO] [✔] AppImage can run on Ubuntu cosmic
appimagelint.glibc_abi_check[21818] [INFO] [✔] AppImage can run on Ubuntu disco
appimagelint.cli[21818] [INFO] Running check "GNU libstdc++ ABI check"
appimagelint.glibcxx_abi_check[21818] [INFO] detected required version for runtime: <none>
appimagelint.glibcxx_abi_check[21818] [INFO] detected required version for payload: 3.4.9
appimagelint.glibcxx_abi_check[21818] [INFO] [✔] AppImage can run on Debian oldstable (jessie)
appimagelint.glibcxx_abi_check[21818] [INFO] [✔] AppImage can run on Debian stable (stretch)
appimagelint.glibcxx_abi_check[21818] [INFO] [✔] AppImage can run on Debian testing (buster)
appimagelint.glibcxx_abi_check[21818] [INFO] [✔] AppImage can run on Debian unstable (sid)
appimagelint.glibcxx_abi_check[21818] [INFO] [✔] AppImage can run on Ubuntu trusty
appimagelint.glibcxx_abi_check[21818] [INFO] [✔] AppImage can run on Ubuntu xenial
appimagelint.glibcxx_abi_check[21818] [INFO] [✔] AppImage can run on Ubuntu bionic
appimagelint.glibcxx_abi_check[21818] [INFO] [✔] AppImage can run on Ubuntu cosmic
appimagelint.glibcxx_abi_check[21818] [INFO] [✔] AppImage can run on Ubuntu disco
```

I'll try the AppImage in a VM later to see if my tool is right. But so far, so good!

**Comment by TheAssassin** [comment]
libc/libstdc++ wise it seems to work. However, you ship kerberos libraries (`libkrb5`, and the version on trusty is seemingly too old (requires the version `KEYUTILS_1.5`, but there's only `KEYUTILS_1.4`).

I guess it's a good idea to extend my tool to also check for these dependencies.

**Comment by jonaski** [comment]
Thanks for the help. I'm not so worried about not supporting older distros than xenial. I think it should run on most of the distros that are still supported.

I actually have the screenshot problem on different machines, the test I did was on my own my machine. Previously I stripped the screenshot out of the appdata file. It works fine on the same machine if I run appstream-util to verify the file outside of appimage deploy. So I don't understand why I get the SSL handshake failure as the URLs work.
On centos it wont run at all (incompatible symbols) that looks to be glib related, so I just remove the appdata file for now.
